pkg/tiller/release_update.go

```diff
@@ -71,13 +71,17 @@ func (s *ReleaseServer) prepareUpdate(req *services.UpdateReleaseRequest) (*rele
 
 	// finds the deployed release with the given name
 	currentRelease, err := s.env.Releases.Deployed(req.Name)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	// If new values were not supplied in the upgrade, re-use the existing values.
-	if err := s.reuseValues(req, currentRelease); err != nil {
-		return nil, nil, err
+	if err == nil {
+		// If new values were not supplied in the upgrade, re-use the existing values.
+		if err := s.reuseValues(req, currentRelease); err != nil {
+			return nil, nil, err
+		}
+	} else {
+		// try to find a failed release we can upgrade
```

Comment:
I'm sorry, but as I mentioned before, we don't want to be upgrading from a failed deploy. We want to upgrade from a completely clean state, as if this was an initial release and there were no deploys released at all. Otherwise we're just reverting back to old behaviour prior to the old PR and we're back to square one.

Reply:
@bacongobbler There's a whole discussion thread about this. tl;dr upgrading from a failed state is a must-be for CI/CD environments - we need idempotent builds.

- https://github.com/kubernetes/helm/issues/3353#issuecomment-359594058
- https://github.com/kubernetes/helm/issues/3353#issuecomment-360374948

Reply:
@bacongobbler I guess you can look at the issue from another angle. When we are installing the first version of the chart and the fails, why do register it at all. If we removed all traces of that deploy then we wouldn't run into this issue.
On the other hand we are loosing the track record that somebody actually tried to deploy something that failed.

Reply:
@Nowaker and @sorenmat : as far as I undertand it in https://github.com/kubernetes/helm/issues/3353#issuecomment-358367529 @bacongobbler suggested to not update from a failed release (which causes diffing with it and in turn various 
other problems) but instead ignore the failed release(s) completely and thread the upgrade like a fresh install (meaning no diffing or reuse of old values). This would keep the release history intact and not require to delete a release manually that has only failed deployments.
Sorry for the interference but I have a feeling the input form @bacongobbler was overlooked and we are digging trenches for nothing.

Reply:
Oh, gotcha. I'm not an opponent or proponent of a particular implementation (I know nothing about Helm's internals). We just want a way to install/upgrade a chart, no matter what state it is now (if any).